### PR TITLE
fix: use multipart upload method to put files larger than 5Gi to OSS. Fixes #12877

### LINKED ISSUE #12877: OSS client  PutObject not support file larger than 5Gi
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issue exists when I tested with `:latest`
- [X] I have searched existing issues and could not find a match for this bug
- [X] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened/what did you expect to happen?

When I tried to put a huge output file larger than 5Gi to aliyun OSS, I received such error messages in wait container:
```
time="2024-04-03T10:10:37.725Z" level=info msg="OSS Save path: /tmp/argo/outputs/artifacts/out.tgz, key: argo-workflows/test/big"
time="2024-04-03T10:10:37.804Z" level=warning msg="failed to put file: oss: service returned error: StatusCode=400, ErrorCode=InvalidArgument, ErrorMessage=\"Post body size must be less than 5G.\", RequestId=660D2B1DAFD89A3732CCE935, Ec=0017-00000102"
time="2024-04-03T10:10:37.804Z" level=warning msg="Non-transient error: oss: service returned error: StatusCode=400, ErrorCode=InvalidArgument, ErrorMessage=\"Post body size must be less than 5G.\", RequestId=660D2B1DAFD89A3732CCE935, Ec=0017-00000102"
time="2024-04-03T10:10:37.804Z" level=info msg="Save artifact" artifactName=out duration=79.453168ms error="oss: service returned error: StatusCode=400, ErrorCode=InvalidArgument, ErrorMessage=\"Post body size must be less than 5G.\", RequestId=660D2B1DAFD89A3732CCE935, Ec=0017-00000102" key=argo-workflows/test/big
time="2024-04-03T10:10:37.804Z" level=error msg="executor error: oss: service returned error: StatusCode=400, ErrorCode=InvalidArgument, ErrorMessage=\"Post body size must be less than 5G.\", RequestId=660D2B1DAFD89A3732CCE935, Ec=0017-00000102"
```

### Version

latest

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: artifact-
spec:
  entrypoint: main
  templates:
    - name: main
      container:
        image: alpine:latest
        command:
          - sh
          - -c
        args:
          - |
            mkdir -p /out
            dd if=/dev/random of=/out/testfile.txt bs=20M count=1024
            echo "created files!"
      outputs:
        artifacts:
          - name: out
            path: /out/testfile.txt
            oss:
              endpoint: http://oss-cn-zhangjiakou-internal.aliyuncs.com
              bucket: bucket
              key: argo-workflows/test/bigfile
              accessKeySecret:
                name: my-argo-workflow-credentials
                key: accessKey
              secretKeySecret:
                name: my-argo-workflow-credentials
                key: secretKey
```

### Logs from the workflow controller

```text
kubectl logs -n argo deploy/workflow-controller | grep ${workflow}
```

### Logs from in your workflow's wait container

```text
kubectl logs -n argo -c wait -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded
```

### PR DESCRIPTION
**Pull request by AlbeeSo**
Fixes #12877

### Motivation

https://github.com/argoproj/argo-workflows/blob/0b625a61a3473f9045d35a6bc0ae8a99db73e2ca/workflow/artifacts/oss/oss.go#L342
OSS SDK client offers several functions help upload local files to server. And PutObjectFromFile uses Simple Upload which usually causes (or always from my experience) InvalidArgument or EntityTooLarge error when uploading files larger than 5Gi. 
More info plz see: https://www.alibabacloud.com/help/en/oss/you-cannot-upload-large-objects-by-using-simple-upload?spm=a3c0i.23458820.2359477120.55.7ef56e9boHbYeC
Multipart Upload is the recommended way.

### Modifications

When trying to save files in OSS by Save Func of ArtifactDriver Interface:
- Still use Simple Upload method if file to saved is not larger than 5Gi.
- Or else split it into `size of capacity / 5Gi +1` parts and upload by Multipart Upload method.

### Verification

Plz test with a simple workflow like
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: artifact-
spec:
  entrypoint: main
  templates:
    - name: main
      container:
        image: alpine:latest
        command:
          - sh
          - -c
        args:
          - |
            mkdir -p /out
            dd if=/dev/random of=/out/testfile.txt bs=20M count=1024
            echo "created files!"
      outputs:
        artifacts:
          - name: out
            path: /out/testfile.txt
            oss:
              endpoint: http://oss-cn-zhangjiakou-internal.aliyuncs.com
              bucket: bucket
              key: argo-workflows/test/bigfile
              accessKeySecret:
                name: my-argo-workflow-credentials
                key: accessKey
              secretKeySecret:
                name: my-argo-workflow-credentials
                key: secretKey
```

### CODE REVIEW

File: workflow/artifacts/oss/oss.go [hunk 1]
```diff
@@ -339,7 +341,51 @@ func isTransientOSSErr(err error) bool {
...
+	fStat, err := os.Stat(path)
+	if err != nil {
+		return err
+	}
+	// Determine upload method based on file size.
+	if fStat.Size() <= maxObjectSize {
+		log.Info("OSS Simple Uploading")
+		return bucket.PutObjectFromFile(objectName, path)
+	}
+	log.Info("OSS Multipart Uploading")
```
Comment (shuangkun) [inline workflow/artifacts/oss/oss.go:353]: can make the Multipart uploading into a separate function.

File: workflow/artifacts/oss/oss.go [hunk 2]
```diff
@@ -339,7 +341,51 @@ func isTransientOSSErr(err error) bool {
...
-	return bucket.PutObjectFromFile(objectName, path)
+	fStat, err := os.Stat(path)
+	if err != nil {
+		return err
+	}
+	// Determine upload method based on file size.
+	if fStat.Size() <= maxObjectSize {
+		log.Info("OSS Simple Uploading")
+		return bucket.PutObjectFromFile(objectName, path)
+	}
```
Comment (shuangkun) [inline workflow/artifacts/oss/oss.go:352]: Is there any reference？

**Review by shuangkun — COMMENTED** [review]
Some comments here。

File: workflow/artifacts/oss/oss.go [hunk 3]
```diff
@@ -339,7 +341,51 @@ func isTransientOSSErr(err error) bool {
...
 func putFile(bucket *oss.Bucket, objectName, path string) error {
 	log.Debugf("putFile from %s to %s", path, objectName)
-	return bucket.PutObjectFromFile(objectName, path)
+	fStat, err := os.Stat(path)
+	if err != nil {
+		return err
+	}
+	// Determine upload method based on file size.
+	if fStat.Size() <= maxObjectSize {
+		log.Info("OSS Simple Uploading")
```
Comment (shuangkun) [inline workflow/artifacts/oss/oss.go:350]: Can change simple as a specific level  https://kubernetes.io/docs/contribute/style/style-guide/#avoid-words-that-assume-a-specific-level-of-understanding

Comment (AlbeeSo) [reply workflow/artifacts/oss/oss.go]: Hi, I referred to the OSS public documentation and use SimpleUpload for consistency, plz see: https://www.alibabacloud.com/help/en/oss/user-guide/simple-upload?spm=a2c63.p38356.0.0.2c072398fh5k3W#33e74f4083r6t

**Comment by AlbeeSo** [comment]
> Some comments here。

Done, PTAL.  :)

**Review by shuangkun — APPROVED** [review]

File: workflow/artifacts/oss/oss.go (hunk 3 above)
Comment (agilgur5) [reply workflow/artifacts/oss/oss.go]: In this case this a name or term, so it is appropriate

**Review by terrytangyuan — APPROVED** [review]
Thanks!

File: workflow/artifacts/oss/oss.go [hunk 4]
```diff
@@ -337,9 +339,66 @@ func isTransientOSSErr(err error) bool {
...
 }
 
+// OSS simple upload code reference: https://www.alibabacloud.com/help/en/oss/user-guide/simple-upload?spm=a2c63.p38356.0.0.2c072398fh5k3W#section-ym8-svm-rmu
+func simpleUpload(bucket *oss.Bucket, objectName, path string) error {
+	log.Info("OSS Simple Uploading")
+	return bucket.PutObjectFromFile(objectName, path)
+}
+
+// OSS multipart upload code reference: https://www.alibabacloud.com/help/en/oss/user-guide/multipart-upload?spm=a2c63.p38356.0.0.4ebe423fzsaPiN#section-trz-mpy-tes
+func multipartUpload(bucket *oss.Bucket, objectName, path string, objectSize int64) error {
```
Comment (agilgur5) [inline workflow/artifacts/oss/oss.go:349]: I know this is mostly taken from the Alibaba docs, but I wonder if the chunking / `UploadPart` section could be parallelized?

**Comment by agilgur5** [comment]
Backported cleanly into `release-3.5` as 1c1f43313

**Pull request merged**